uid-mux/src/yamux.rs

```diff
@@ -0,0 +1,440 @@
+
+            Poll::Pending
+        }
+    }
+}
+
+/// A yamux control handle.
+#[derive(Debug, Clone)]
+pub struct YamuxCtrl {
+    role: Role,
+    queue: Arc<Mutex<Queue>>,
+    close_notify: Arc<Notify>,
+    shutdown_notify: Arc<AtomicBool>,
+}
+
+impl YamuxCtrl {
+    /// Close the yamux connection.
+    pub fn close(&self) {
+        self.shutdown_notify.store(true, Ordering::Relaxed);
+        self.wake_conn();
+    }
+
+    /// Wakes up the connection.
+    fn wake_conn(&self) {
+        self.queue.lock().unwrap().waker.wake_by_ref();
+    }
+}
+
+#[async_trait]
+impl<Id> UidMux<Id> for YamuxCtrl
+where
+    Id: fmt::Debug + AsRef<[u8]> + Sync,
+{
+    type Stream = Stream;
+    type Error = std::io::Error;
+
+    #[cfg_attr(
+        feature = "tracing",
+        tracing::instrument(
+            fields(role = %self.role, id = hex::encode(id)),
+            skip_all,
+            err
+        )
+    )]
+    async fn open(&self, id: &Id) -> Result<Self::Stream, Self::Error> {
+        let internal_id = InternalId::new(id.as_ref());
+
+        debug!("opening stream: {}", internal_id);
+
+        let receiver = {
+            let mut queue = self.queue.lock().unwrap();
+            if let Some(stream) = queue.ready.remove(&internal_id) {
+                trace!("stream already opened");
+                return Ok(stream);
+            }
+
+            let (sender, receiver) = oneshot::channel();
+
+            // Insert the oneshot into the queue.
+            queue.waiting.insert(internal_id, sender);
+            // Wake up the connection.
+            queue.waker.wake_by_ref();
+
+            trace!("waiting for stream");
+
+            receiver
+        };
+
+        futures::select! {
+            stream = receiver.fuse() =>
+                stream
+                    .inspect(|_| debug!("caller received stream"))
+                    .inspect_err(|_| error!("connection cancelled stream"))
+                    .map_err(|_| {
+                    std::io::Error::other(format!("connection cancelled stream"))
+                }),
+            _ = self.close_notify.notified().fuse() => {
+                error!("connection closed before stream opened");
+                Err(std::io::ErrorKind::ConnectionAborted.into())
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use futures::{AsyncReadExt, AsyncWriteExt};
+    use tokio::io::duplex;
+    use tokio_util::compat::TokioAsyncReadCompatExt;
+
+    #[tokio::test]
+    async fn test_yamux() {
+        let (client_io, server_io) = duplex(1024);
+        let client = Yamux::new(client_io.compat(), Config::default(), Mode::Client);
+        let server = Yamux::new(server_io.compat(), Config::default(), Mode::Server);
+
+        let client_ctrl = client.control();
+        let server_ctrl = server.control();
+
+        let conn_task = tokio::spawn(async {
+            futures::try_join!(client.into_future(), server.into_future()).unwrap();
```

Comment:
Would it be possible to test 2 situations disctinctly:
- the future resolves because of ctrl.close()
- the future resolves because the remote called its ctrl.close()

right now it's not obvious why it resolves